oak_restricted_kernel/src/lib.rs

```diff
@@ -246,46 +246,46 @@ pub fn start_kernel(info: &BootParams) -> ! {
 
     let stage0_dice_data = {
         let dice_memory_slice = {
-            let e820_dice_data_entry = info
-                .e820_table()
-                .iter()
-                .find(|e| e.entry_type() == Some(oak_linux_boot_params::E820EntryType::DiceData))
-                .expect("failed to find dice data");
-
-            let phys_start_addr = PhysAddr::new_truncate(
-                e820_dice_data_entry
-                    .addr()
-                    .try_into()
-                    .expect("couldn't convert usize to u64"),
-            );
-
-            // Validate that the dice data mem address matches the kernel args if present
-            if let Some(arg) = kernel_args.get(&alloc::format!(
-                "--{}",
-                oak_dice::evidence::DICE_DATA_CMDLINE_PARAM
-            )) {
+            let dice_data_phys_addr = {
+                let arg = kernel_args
+                    .get(&alloc::format!(
+                        "--{}",
+                        oak_dice::evidence::DICE_DATA_CMDLINE_PARAM
+                    ))
+                    .expect("no dice data address supplied in the kernel args");
                 let parsed_arg = u64::from_str_radix(
                     arg.strip_prefix("0x")
                         .expect("failed stripping the hex prefix"),
                     16,
                 )
                 .expect("couldn't parse address as a hex number");
-                if parsed_arg != phys_start_addr.as_u64() {
-                    panic!("inconsistent dice data addresses supplied in the E820 table and kernel args")
-                }
-            }
+                PhysAddr::new(parsed_arg)
+            };
+
+            // Ensure that the dice data is stored within reserved memory.
+            assert!(info.e820_table().iter().any(|e| {
+                let dice_data_fully_contained_in_segment = e.addr() as u64
+                    <= dice_data_phys_addr.as_u64()
+                    && (dice_data_phys_addr.as_u64()
+                        + core::mem::size_of::<oak_dice::evidence::Stage0DiceData>() as u64)
+                        <= (e.addr() + e.size()) as u64;
+
+                e.entry_type().expect("failed to get type")
+                    == oak_linux_boot_params::E820EntryType::RESERVED
+                    && dice_data_fully_contained_in_segment
+            }));
```

Comment:
would fail if the dice data is stored across two directly adjacent reserved memory segments. seems fine though.